src/bebb/main.mo

```diff
@@ -631,6 +650,44 @@ actor {
     };
   };
 
+  /**
+   * Helper functions to match filters against all Entities
+  */
+  private func matchEntities(caller : Principal, filterCriteria : [Entity.EntityFilterCriterion]) : ?[Entity.Entity] {
+    if (filterCriteria.size() < 1) {
+      return null;
+    };
+    var foundMatches : Iter.Iter<Entity.Entity> = entitiesStorage.vals();
+    for (filterCriterion in filterCriteria.vals()) {
+      foundMatches := filterEntities(caller, filterCriterion, foundMatches)
+    };
+    return ?Iter.toArray<Entity.Entity>(foundMatches);
+  };
+
+  private func filterEntities(caller : Principal, filterCriterion : Entity.EntityFilterCriterion, entitiesToFilter : Iter.Iter<Entity.Entity>) : Iter.Iter<Entity.Entity> {
+    let buffer = Buffer.Buffer<Entity.Entity>(8);
+    Iter.iterate(entitiesToFilter, func(entity : Entity.Entity, ix : Nat) {
+      switch (applyFilterToEntity(caller, filterCriterion, entity)) {
+        case (true) { buffer.add(entity) };
+        case _ {};
+      };
+    });
+    return buffer.vals();
+  };
+
+  private func applyFilterToEntity(caller : Principal, filterCriterion : Entity.EntityFilterCriterion, entity : Entity.Entity) : Bool {
+    switch (filterCriterion.criterionKey) {
+      case ("externalId") {
```

Comment:
Do we need to be explicit about the externID? Or can it be any general string they pass in and as long as that field exists then it can compare against it?

Reply:
that's an interesting idea and worth adding. Once we have more filter criteria (and especially ones that wouldn't match against entitySpecificFields), we can reconsider the implementation